go/vt/vttablet/tabletmanager/vreplication/vreplicator.go

```diff
@@ -558,3 +570,298 @@ func recalculatePKColsInfoByColumnNames(uniqueKeyColumnNames []string, colInfos
+			sqltypes.StringBindVariable(dbconfigs.GlobalDBConfigs.Filtered.User), sqltypes.StringBindVariable(alterStmt))
+		if err != nil {
+			return err
+		}
+		killdbc := vr.vre.dbClientFactoryDba()
+		if err := killdbc.Connect(); err != nil {
+			return fmt.Errorf("unable to connect to the database when attempting to kill %q: %v", alterStmt, err)
+		}
+		defer killdbc.Close()
+		res, err := killdbc.ExecuteFetch(getID, 100)
+		if err != nil {
+			return err
+		}
+		if len(res.Rows) == 0 {
+			return nil // Not running
+		}
+		if len(res.Rows) > 1 {
+			return fmt.Errorf("unexpected number of rows returned (%d) when attempting to kill %q", len(res.Rows), alterStmt)
+		}
+		id, err := res.Rows[0][0].ToInt64()
+		if err != nil {
+			return fmt.Errorf("unexpected connection ID value returned (%+v) when attempting to kill %q", res.Rows[0][0], alterStmt)
+		}
+		_, err = killdbc.ExecuteFetch(fmt.Sprintf("kill %d", id), 1)
+		return err
+	}
+	go func() {
+		select {
+		case <-ctx.Done():
+			log.Infof("Copy of %q table stopped when performing the following post copy action for the %q VReplication workflow: %+v",
+				tableName, vr.WorkflowName, action)
+			if err := killAlterStmt(action.Task); err != nil {
+				log.Errorf("Failed to kill post copy ALTER statement for the %q VReplication workflow: %v", vr.WorkflowName, err)
+			}
+			return
+		case <-done:
+			// We're done, so no longer need to listen for cancellation.
+			return
+		}
+	}()
+
+	for _, row := range qr.Named().Rows {
+		select {
+		case <-ctx.Done():
+			return vterrors.Errorf(vtrpcpb.Code_CANCELED, "context has expired")
+		default:
+		}
+		action = PostCopyAction{}
+		acv, err := row["action"].ToBytes()
+		if err != nil {
+			return err
+		}
+		if err := json.Unmarshal(acv, &action); err != nil {
+			return err
+		}
+
+		switch action.Type {
+		case PostCopyActionSQL:
+			log.Infof("Executing post copy SQL action for the %q table in the %q VReplication workflow: %s",
+				tableName, vr.WorkflowName, action.Task)
+			// This will return an io.EOF / MySQL CRServerLost (errno 2013)
+			// error if it is killed by the monitoring goroutine.
+			if _, err := dbClient.ExecuteFetch(action.Task, -1); err != nil {
+				failedAlterErr := err
+				// It's possible that we previously executed the ALTER but
+				// the subsequent DELETE of the post_copy_action record failed.
+				// For example, the context could be cancelled in-between.
+				// It's also possible that the user has modified the schema on
+				// the target side.
+				// If we get a duplicate key/index error then let's see if the
+				// index definitions that we would have added already exist in
+				// the table schema and if so move forward and delete the
+				// post_copy_action record.
+				if sqlErr, ok := err.(*mysql.SQLError); ok && sqlErr.Number() == mysql.ERDupKeyName {
+					stmt, err := sqlparser.ParseStrictDDL(action.Task)
```

Comment:
ah, OK, I think this answers my earlier question about idempotency. Please confirm.

Reply:
Yep!